### PR TITLE
Integrate etcd into the test cases themselves

### PR DESCRIPTION
**Pull request by smarterclayton**
@liggitt @deads2k

### CODE REVIEW

File: hack/test-integration.sh [hunk 1]
```diff
@@ -93,11 +83,6 @@ function exectest() {
 		os::text::print_red "failed  $1"
 		echo "${out}"
 
-		# dump etcd for failing test
-		echo "[INFO] Dumping etcd contents to ${ARTIFACT_DIR}/$1-etcd_dump.json"
```
Comment (deads2k) [inline hack/test-integration.sh]: This helped me during the rebase.  I'd be sad to see it go.

**Comment by deads2k** [comment]
I tried this in https://github.com/openshift/origin/pull/6474 and it didn't really buy me anything.  Why are you doing this?

**Comment by deads2k** [comment]
See: https://github.com/openshift/origin/pull/6474#issuecomment-166737869

**Comment by smarterclayton** [comment]
This isn't speed, this is simplicity.  This is about reducing the number of components that cause problems.  Having to do builds before integration is not what integration is about, nor is the external etcd really valuable on its own.

I'd be fine adding a "dump etcd" env var to the test case that writes a dump of the current state of etcd at the end of the test to the logs with a unique name based on the test.

**Comment by deads2k** [comment]
Speaking of simplicity, do you have an explanation of how everything is parallelized?  There is an iops storm here: https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/10884/artifact/origin/artifacts/test-integration/logs/ that is causing etcd to fail, but I don't know what's in parallel with it.

**Comment by smarterclayton** [comment]
Test integration is not parallel (only unit and cmd are, and only with each
other).

**Comment by deads2k** [comment]
Also, if this runs slower, I'm against doing it.  Running against external etcd saved significant time in my tests against serial execution and wasn't any harder to reason about.

**Comment by smarterclayton** [comment]
The performance difference is ~5-10 seconds across all tests.

**Comment by smarterclayton** [comment]
Which is basically background noise.  This also removes the need for
deleteAllKeys, and any potential races between controllers, and indexes
that are not independent, and a pre build step, and any potential for
racing if we do parallelize in the future.

**Comment by deads2k** [comment]
[test] for timing

**Comment by smarterclayton** [comment]
1053 for this vs 1108 for another recent run, so basically same speed or margin of error.

However, you can see the one failure which is a race / flake.  I need to debug more, but this could be stressing the system enough so that we're seeing a recreateable scenario for the previous etcd failure here.

**Comment by deads2k** [comment]
With this split, you could check for "has no leader" and re-run until we're able to fix it.  I don't have any good ideas for fix it.

**Comment by smarterclayton** [comment]
I think this is the same problem as the healthz timeout in the apiserver in
our tests.  A retry in the master on etcd connection for this specific
condition (wait until leader convergence) seems reasonable.  We used to
retry until we could talk to the etcd server - did that code change / move
recently?

**Comment by deads2k** [comment]
Yeah, that code died.  `GetAndTestEtcdClient` or some such.  But that's not what you're actually getting.  That error pops up even on an etcd that's been around for a while and successfully responded to other  requests.  I'd bet on the suddenly slow disk problem.  I think they have some metrics gathered on it, but I haven't pursued it yet.

**Comment by smarterclayton** [comment]
magic [test] genies, give me some more data points.

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
Fixed the issue with races for tests - [merge]

File: test/integration/api_paths_test.go [hunk 2]
```diff
@@ -27,6 +27,7 @@ func TestRootAPIPaths(t *testing.T) {
 		"/logs/": http.StatusNotFound,
 	}
 
+	defer testutil.RequireEtcd(t).Terminate(t)
```
Comment (liggitt) [inline test/integration/api_paths_test.go]: Does this evaluate the first function immediately, and the second function deferred?

Comment (smarterclayton) [inline test/integration/api_paths_test.go]: Yes

Comment (smarterclayton) [inline test/integration/api_paths_test.go]: I wouldn't recommend this in normal code but in test code it's just there
to be good citizens

**Pull request merged**